### ISSUE TITLE
External user permissions

### ISSUE THREAD

**Issue by muellermartin** [go-gitea/gitea#4334]
GitLab provides a special type of local users called "external users". The [current version of GitLab's documentation](https://docs.gitlab.com/ee/user/permissions.html#external-users-permissions) descibes this role as follows:

> In cases where it is desired that a user has access only to some internal or private projects, there is the option of creating External Users. This feature may be useful when for example a contractor is working on a given project and should only have access to that project.
> 
> External users can only access projects to which they are explicitly granted access, thus hiding all other internal or private ones from them. Access can be granted by adding the user as member to the project or group.
> 
> They will, like usual users, receive a role in the project or group with all the abilities that are mentioned in the table above. They cannot however create groups or projects, and they have the same access as logged out users in all other cases.

This type of user is particulary useful when the Gitea instance is private and a guest/external user should access only specific repositories while the rest is hidden. Currently (as of version 1.4.3) local users can see other repositories when they use the "Explore" functionality even if they are not part of any organization or added as collaborator to these repositories.

Remaining tasks:
* [x] Filter results when exploring users
* [x] Filter /org/orgname/dashboard content
* [ ] Ability for external users to create organizations and repositories (separate option for count limit)
* [x] LDAP option to automatically set restricted user

**Comment by mnsh** [comment]
Such a feature would be very useful to any organization using gitea internally with the need to give customers access to their projects.

I propose introducing a new user flag called "IsRestricted", presented as another checkbox in the user admin pages.

A restricted user's view would be limited to:
- Repos accessible by the user's teams as well as repos where the user is a collaborator
- Orgs where the user is member of a team
- Users that are members of teams in those orgs

To put it another way, a restricted user would simply not see or be able to access anything public.

**Comment by brunoais** [comment]
How close is this to be done?

**Comment by mnsh** [comment]
It is done (#6274)
